source/base/index_set.cc

```diff
@@ -257,6 +257,144 @@ IndexSet::get_view(const size_type begin, const size_type end) const
+      // is recorded in the mask's ranges.
+      if ((own_it->begin <= mask_it->begin) && (own_it->end <= mask_it->end))
+        {
+          new_ranges.emplace_back(mask_it->begin - mask_it->nth_index_in_set,
+                                  own_it->end - mask_it->nth_index_in_set);
+        }
+      else
+        // Case 2:our interval overlaps the tail end of the other interval
+        if ((mask_it->begin <= own_it->begin) && (mask_it->end <= own_it->end))
+          {
+            const size_type offset_within_mask_interval =
+              own_it->begin - mask_it->begin;
+            new_ranges.emplace_back(mask_it->nth_index_in_set +
+                                      offset_within_mask_interval,
+                                    mask_it->nth_index_in_set +
+                                      (mask_it->end - mask_it->begin));
+          }
+        else
+          // Case 3: Our own interval completely encloses the other interval
+          if ((own_it->begin <= mask_it->begin) &&
+              (own_it->end >= mask_it->end))
+            {
+              new_ranges.emplace_back(mask_it->begin -
+                                        mask_it->nth_index_in_set,
+                                      mask_it->end - mask_it->nth_index_in_set);
+            }
+          else
+            // Case 3: The other interval completely encloses our own interval
+            if ((mask_it->begin <= own_it->begin) &&
+                (mask_it->end >= own_it->end))
+              {
+                const size_type offset_within_mask_interval =
+                  own_it->begin - mask_it->begin;
+                new_ranges.emplace_back(mask_it->nth_index_in_set +
+                                          offset_within_mask_interval,
+                                        mask_it->nth_index_in_set +
+                                          offset_within_mask_interval +
+                                          (own_it->end - own_it->begin));
+              }
+            else
+              Assert(false, ExcInternalError());
+
+      // We considered the overlap of these two intervals. It may of course
+      // be that one of them overlaps with another one, but that can only
+      // be the case for the interval that extends further to the right. So
+      // we can safely move on from the interval that terminates earlier:
+      if (own_it->end < mask_it->end)
+        ++own_it;
+      else if (mask_it->end < own_it->end)
+        ++mask_it;
+      else
+        {
+          // The intervals ended at the same point. We can move on from both.
+          // (The algorithm would also work if we only moved on from one,
+          // but we can micro-optimize here without too much effort.)
+          ++own_it;
+          ++mask_it;
+        }
+    }
+
+  // Now turn the ranges of overlap we have accumulated into an IndexSet in
+  // its own right:
+  IndexSet result(size());
```

Comment:
Didn't you want to make the size equal the number of entries in the mask: The documentation states `then the result will be an index set for a space of size 50 that contains those indices that [...]`. Overall, I am not sure yet how you want to achieve this in a parallel setting. If I read the documentation correctly, I need to be sure that the mask can be defined for the entirety of the indices on **all processes** in the user code, and only afterwards call this here. Is this understanding correct?

Reply:
Oh yes, excellent catch! This should have been 
```suggestion
  IndexSet result(mask.n_elements());
```
I made the fix, and also added assertions in this regard to the three tests added here.